src/wallet/wallet.cpp

```diff
@@ -3999,52 +3999,41 @@ bool CWallet::ApplyMigrationData(MigrationData& data, bilingual_str& error)
 
     // Check the address book data in the same way we did for transactions
     std::vector<CTxDestination> dests_to_delete;
-    for (const auto& addr_pair : m_address_book) {
-        // Labels applied to receiving addresses should go based on IsMine
-        if (addr_pair.second.purpose == AddressPurpose::RECEIVE) {
-            if (!IsMine(addr_pair.first)) {
-                // Check the address book data is the watchonly wallet's
-                if (data.watchonly_wallet) {
-                    LOCK(data.watchonly_wallet->cs_wallet);
-                    if (data.watchonly_wallet->IsMine(addr_pair.first)) {
-                        // Add to the watchonly. Copy the entire address book entry
-                        data.watchonly_wallet->m_address_book[addr_pair.first] = addr_pair.second;
-                        dests_to_delete.push_back(addr_pair.first);
-                        continue;
-                    }
-                }
-                if (data.solvable_wallet) {
-                    LOCK(data.solvable_wallet->cs_wallet);
-                    if (data.solvable_wallet->IsMine(addr_pair.first)) {
-                        // Add to the solvable. Copy the entire address book entry
-                        data.solvable_wallet->m_address_book[addr_pair.first] = addr_pair.second;
-                        dests_to_delete.push_back(addr_pair.first);
-                        continue;
-                    }
-                }
+    for (const auto& [dest, record] : m_address_book) {
+        // Ensure "receive" entries that are no longer part of the original wallet are transferred to another wallet
+        // Entries for everything else ("send") will be cloned to all wallets.
+        bool require_transfer = record.purpose == AddressPurpose::RECEIVE && !IsMine(dest);
+        bool copied = false;
+        for (auto& wallet : {data.watchonly_wallet, data.solvable_wallet}) {
+            if (!wallet) continue;
+
+            LOCK(wallet->cs_wallet);
+            if (require_transfer && !wallet->IsMine(dest)) continue;
+
+            // Copy the entire address book entry
+            wallet->m_address_book[dest] = record;
+
+            copied = true;
+            // Only delete 'receive' records that are no longer part of the original wallet
+            if (require_transfer) {
+                dests_to_delete.push_back(dest);
+                break;
```

Comment:
> If the destination requires_transfer but isn't ours, go to the next wallet. If this happens for both wallets, copy will never be set and we will fail

If the destination requires transfer (a "receive" addr that is no longer part of the main wallet), the process checks to which wallet the entry belongs to. And, when the entry does not belong to any of the created wallets, it means that the key/script-to-descriptor migration process failed. This is because "receive" entries must always be tracked by the wallet.

> If it requires transfer and does belong to the current wallet in the loop, continue and set copy = true, and add the destination to dests_to_delete. Do not check the next wallet. I'm assuming the assumption here is it can only belong to one of the wallets, and if it does end up happening to belong to both wallets this is not an issue?

Yes, thats the existing assumption. If you check the previous code, will see that I did not change this part of the flow on purpose. This is what we had before:

```c++
if (purpose == "RECEIVE" && not_mine_anymore) {

    // check if it is part of the watch-only wallet
    if (watch_only_wallet.IsMine(dest)) {
        store_address(watch_only_wallet, dest);
        add_to_pending_to_delete(dest);
        continue; // --> this skips the next wallet.
    } 

    // check if it is part of the solvables wallet
    if (solvables_wallet.IsMine(dest)) {
        store_address(solvables_wallet, dest);
        add_to_pending_to_delete(dest);
        continue; // --> skips any following-up check
    } 

    // We only reach here if the entry wasn't added to any wallet
    error out...
}

```

The only diff between the new code and the previous one is that we had a `continue` instead of a `break` statement.


 > Might be worth adding a comment explaining why it's okay to break early? That's the part that kinda threw me.

We should double-check if we could ever have the same no-longer ours "receive" entry in both wallets (the watch-only and the solvables ones). If we find a way to achieve this, then the bug predates this PR.